bfv/bfv_test.go

```diff
@@ -227,413 +228,607 @@ func testEncoder(testctx *testContext, t *testing.T) {
 				coeffsInt[i] = int64(c)
 			}
 		}
-		plaintext := NewPlaintext(testctx.params)
-		testctx.encoder.EncodeInt(coeffsInt, plaintext)
-		coeffsTest := testctx.encoder.DecodeIntNew(plaintext)
+		plaintext := NewPlaintext(tc.params)
+		tc.encoder.EncodeInt(coeffsInt, plaintext)
+		coeffsTest := tc.encoder.DecodeIntNew(plaintext)
 
 		require.True(t, utils.EqualSliceInt64(coeffsInt, coeffsTest))
 	})
 
-	t.Run(testString("Encoder/Encode&Decode/PlaintextMul", testctx.params), func(t *testing.T) {
-		values, plaintext := newTestVectorsMul(testctx, t)
-		verifyTestVectors(testctx, nil, values, plaintext, t)
+	t.Run(testString("Encoder/Encode&Decode/PlaintextMul", tc.params), func(t *testing.T) {
+		values, plaintext := newTestVectorsMul(tc, t)
+		verifyTestVectors(tc, nil, values, plaintext, t)
 	})
 }
 
-func testEvaluator(testctx *testContext, t *testing.T) {
+func testQuantize(tc *testContext, t *testing.T) {
 
-	t.Run(testString("Evaluator/Add/op1=Ciphertext/op2=Ciphertext", testctx.params), func(t *testing.T) {
+	t.Run(testString("QuantizeToLvl", tc.params), func(t *testing.T) {
 
-		values1, _, ciphertext1 := newTestVectorsRingQ(testctx, testctx.encryptorPk, t)
-		values2, _, ciphertext2 := newTestVectorsRingQ(testctx, testctx.encryptorPk, t)
+		values1, _, ciphertext1 := newTestVectorsRingQ(tc, tc.encryptorPk, t)
 
-		testctx.evaluator.Add(ciphertext1, ciphertext2, ciphertext1)
-		testctx.ringT.Add(values1, values2, values1)
+		tc.evaluator.QuantizeToLvl(1, ciphertext1)
+		verifyTestVectors(tc, tc.decryptor, values1, ciphertext1, t)
 
-		verifyTestVectors(testctx, testctx.decryptor, values1, ciphertext1, t)
+		if tc.params.T() != tc.params.RingQ().Modulus[0] { // only happens if T divides Q.
+			tc.evaluator.QuantizeToLvl(0, ciphertext1)
+			verifyTestVectors(tc, tc.decryptor, values1, ciphertext1, t)
+		}
 	})
 
-	t.Run(testString("Evaluator/AddNoMod/op1=Ciphertext/op2=Ciphertext", testctx.params), func(t *testing.T) {
+	t.Run(testString("QuantizeToLvl/Add", tc.params), func(t *testing.T) {
+
+		values1, _, ciphertext1 := newTestVectorsRingQ(tc, tc.encryptorPk, t)
+		values2, _, ciphertext2 := newTestVectorsRingQ(tc, tc.encryptorPk, t)
 
-		values1, _, ciphertext1 := newTestVectorsRingQ(testctx, testctx.encryptorPk, t)
-		values2, _, ciphertext2 := newTestVectorsRingQ(testctx, testctx.encryptorPk, t)
+		tc.evaluator.QuantizeToLvl(1, ciphertext1)
+		tc.evaluator.QuantizeToLvl(1, ciphertext2)
```

Comment:
test should check that the level is indeed 1 after this operation.

Reply:
True

Reply:
Done